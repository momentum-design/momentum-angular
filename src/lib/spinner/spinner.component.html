<div
  *ngIf="isPercentage; else normal"
  [ngClass]="percentageClasses"
  class="md-spinner-progress"
>
  <div class="md-spinner-progress__circle">
    <div class="md-spinner-progress__mask md-spinner-progress__full">
      <div class="md-spinner-progress__fill"></div>
    </div>
    <div class="md-spinner-progress__mask md-spinner-progress__half">
      <div class="md-spinner-progress__fill"></div>
      <div class="md-spinner-progress__fill md-spinner-progress__fix"></div>
    </div>
    <div
      [ngClass]="checkMarkClasses"
      class="md-spinner-progress__inset-circle"
    >
      <div
        *ngIf="size === 36 && showPercentage && !showCheck"
        class="md-spinner-progress__percentage"
      >
        {{ percentage | number: '1.0-0' }}
      </div>
    </div>
  </div>
</div>

<ng-template #normal>
  <i [ngClass]="defaultClasses" class="md-spinner"></i>
</ng-template>
