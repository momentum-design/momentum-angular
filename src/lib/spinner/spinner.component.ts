import { Component, Input, OnInit } from "@angular/core";

@Component({
  selector: 'cui-spinner',
  templateUrl: './spinner.component.html',
  styles: [],
})
export class SpinnerComponent implements OnInit {

  @Input() public className: string;
  @Input() public color: string = 'black';
  @Input() public size: number = 36;
  @Input() public percentage: number;
  @Input() public showCheck: boolean = false;
  @Input() public showPercentage: boolean = false;

  public isPercentage : boolean;

  constructor() { }

  ngOnInit() {
    this.checkPercentage(this.percentage);
  }

  get defaultClasses() {
    return {
      ['cui-spinner--' + this.size]: this.sizeCheck(this.size),
      ['cui-spinner--36']: Number(this.size) === 36,
      ['cui-spinner--' + this.color]: this.color,
      [this.className]: this.className,
    };
  }

  get percentageClasses() {
    return {
      ['cui-spinner-progress--' + this.size]: this.sizeCheck(this.size),
      ['cui-spinner-progress__percentage-' + this.percentage]: this.percentage,
      ['cui-spinner-progress--' + this.color]: this.color,
      [this.className]: this.className,
    };
  }

  get checkMarkClasses() {
    return {
      ['cui-spinner-progress__inset-circle--check']: this.sizeCheck(this.size) && this.showCheck,
      ['cui-spinner-progress__inset-circle--check' + this.percentage]: this.showCheck && Number(this.percentage) === 100,
    };
  }

  private checkPercentage = (percent) => {
    if(!isNaN(percent)){
      this.isPercentage = true;
      this.percentage = Math.round(percent);
    }
    if (Number(this.size) !== 36 && this.showPercentage) {
      console.warn('Spinner: Percentage will not be shown for sizes smaller than 36');
    }
  }

  private sizeCheck = (size) => {
    const sizes = [16, 20, 28, 36];
    if(sizes.includes(Number(size))){
      return true;
    }
    console.warn('Only sizes 16, 20, 28, 36 are acceptable for spinner')
    return false;
  }
}

/**
 * @component loader-spinner
 * @section default
 * @angular
 *
<div>
  <cui-spinner></cui-spinner>
</div>
 */

 /**
 * @component loader-spinner
 * @section determinate
 * @angular
 *
<div>
  <cui-spinner percentage=65></cui-spinner>
</div>
 */


 /**
 * @component loader-spinner
 * @section percentage
 * @angular
 *
<div>
  <cui-spinner percentage=65 showPercentage=true></cui-spinner>
</div>
 */


 /**
 * @component loader-spinner
 * @section check
 * @angular
 *
<div>
  <cui-spinner percentage=100 showCheck=true></cui-spinner>
</div>
 */
