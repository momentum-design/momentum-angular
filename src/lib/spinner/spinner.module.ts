import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SpinnerComponent } from './spinner.component';

@NgModule({
  declarations: [SpinnerComponent],
  imports: [CommonModule],
  exports: [SpinnerComponent],
})
export class SpinnerModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: SpinnerModule,
      providers: [],
    };
  }
}
