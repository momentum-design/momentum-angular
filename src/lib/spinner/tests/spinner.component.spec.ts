import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { SpinnerComponent } from '../spinner.component';

describe('SpinnerComponent', () => {
  let component: SpinnerComponent;
  let fixture: ComponentFixture<SpinnerComponent>;
  let iconNativeElement: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [SpinnerComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SpinnerComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create and match snapshot', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render with blue color class name', () => {
    component.color = 'blue';
    fixture.detectChanges();

    iconNativeElement = fixture.nativeElement;
    const i = iconNativeElement.querySelector('i');

    const expectedIconName = 'cui-spinner--blue';
    expect(i.className).toContain(expectedIconName);
  });

  it('should render with size 28', () => {
    component.size = 28;
    fixture.detectChanges();

    iconNativeElement = fixture.nativeElement;
    const i = iconNativeElement.querySelector('i');

    const expectedIconName = 'cui-spinner--28';
    expect(i.className).toContain(expectedIconName);
  });

  it('should render with className test', () => {
    component.className = 'test';
    fixture.detectChanges();

    iconNativeElement = fixture.nativeElement;
    const i = iconNativeElement.querySelector('i');

    const expectedIconName = 'test';
    expect(i.className).toContain(expectedIconName);
  });

  // it('should show a 65% fill', () => {
  //   component.percentage = 65;
  //   fixture.detectChanges();

  //   iconNativeElement = fixture.nativeElement;
  //   const div = iconNativeElement.querySelector('.cui-spinner-progress');
  //   const expectedIconName = 'cui-spinner-progress__percentage-65';

  //   expect(div.className).toContain(expectedIconName);

  // });
});
