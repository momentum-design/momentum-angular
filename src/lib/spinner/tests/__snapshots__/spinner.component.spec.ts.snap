// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerComponent should create and match snapshot 1`] = `
<cui-spinner
  checkPercentage={[Function Function]}
  color={[Function String]}
  showCheck="false"
  showPercentage="false"
  size={[Function Number]}
  sizeCheck={[Function Function]}
>
  <i
    class="cui-spinner cui-spinner--36 cui-spinner--black"
    ng-reflect-klass="cui-spinner"
    ng-reflect-ng-class="[object Object]"
  />
</cui-spinner>
`;
